### PR TITLE
Dyn 4738 custom color picker

### PR DESCRIPTION
**Pull request by RobertGlobant20**
### Purpose

Replacing the current ColorPicker and ColorPalette used for GroupStyles and for the Color Palette node.
Currently in Dynamo the GroupStyles are using the System.Windows.Forms.ColorDialog control for selecting the color and the Color Palette node is using the WPF Extended Toolkit ColorPicker, now with this change in both places we will be using the CustomColorPicker (based in WPF Extended Toolkit ColorPicker).

Basically I did the next tasks:
- Remove WPF Extended Toolkit 3.0.0 from CoreNodeModelsWpf project.
- Add WPF Extended Toolkit 3.0.0 to DynamoCoreWpf project
- Create a new ColorPicker named CustomColorPicker (based in WPF Extended Toolkit ColorPicker).
- Use the new ColorPicker in Color Palette node.
- Use the new ColorPicker for selecting the color when creating new styles (Preferences -> GroupStyles).

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [X] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

Replacing the current ColorPicker and ColorPalette used for GroupStyles and for the Color Palette node.

### Reviewers

@QilongTang 

### FYIs

@zeusongit

### CODE REVIEW

**Comment by RobertGlobant20** [comment]
This is a GIF showing the behavior of the CustomColorPicker in both places:
CustomColorPicker

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -3641,4 +3641,60 @@ private double GetRelativeLuminance(System.Windows.Media.Color color)
...
+        /// <summary>
+        /// Converts a Color to a SolidColorBrush.
+        /// </summary>
+        /// <returns>
+        /// A converted SolidColorBrush. If the method returns null, the valid null value is used.
+        /// </returns>
+        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
+        {
+            if (value != null)
+                return new SolidColorBrush((Color)value);
```
Comment (zeusongit) [inline src/DynamoCoreWpf/UI/Converters.cs:3680]: So here we do not return `((SolidColorBrush)value).Color` property, like we do in `ConvertBack` below?
https://github.com/DynamoDS/Dynamo/pull/13794/files#diff-1eceea7bbe76f1c17266f4a4eafefa61ccf9c46412dc8c01ea9aa3c5aaef7a6cR3695
 Is that intended?

**Comment by zeusongit** [comment]
From the GIF I see that there is considerable lag when you open the color picker for the first time, is it because you are in Debug, or is it usual?

**Comment by Amoursol** [comment]
@RobertGlobant20 looks like we're missing the "Advanced" section that allows users to select and specify their own color in full RGB space. Not having this would be a pretty significant downgrade on existing ability to select colors. 


image
image

**Comment by Amoursol** [comment]
@RobertGlobant20 we should also update the style of the base node here too, as the border and dropdown are of the old style:

image

**Comment by RobertGlobant20** [comment]
> @RobertGlobant20 looks like we're missing the "Advanced" section that allows users to select and specify their own color in full RGB space. Not having this would be a pretty significant downgrade on existing ability to select colors.
> 
> image image

@Amoursol The advanced section will be part of the next Jira task, right now that section is hidden (we decide it to do it in a different task due that is a complete different window with different style).
Thanks

**Comment by RobertGlobant20** [comment]
> From the GIF I see that there is considerable lag when you open the color picker for the first time, is it because you are in Debug, or is it usual?
@zeusongit well ... my computer (the old one) is extremely slow even loading the Library takes several seconds to show so I expect for normal users should be quicker. Thanks

File: src/DynamoCoreWpf/UI/Converters.cs (hunk 1 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/UI/Converters.cs:3680]: not sure if I misunderstood your question but this is the explanation: the ColorToSolidColorBrushConverter converter converts from Color to SolidColorBrush and convert it back so using this SolidColorBrush((Color)value) will convert Color to SolidColorBrush.
For ConvertBack we receive a SolidColorBrush, so in order to convert it to Color we need to use ((SolidColorBrush)value).Color - first cast to SolidColorBrush and the get the SolidColorBrush.Color.
Let me know if I answered your comment.

Comment (zeusongit) [reply src/DynamoCoreWpf/UI/Converters.cs:3680]: Yes, that answers it, thanks.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 2]
```diff
@@ -3641,4 +3641,60 @@ private double GetRelativeLuminance(System.Windows.Media.Color color)
             return 0.2126 * R + 0.7152 * G + 0.0722 * B;
         }
     }
+
+    /// <summary>
+    /// This converter is used to add margin/Padding from Popup in the CustomColorPicker
```
Comment (QilongTang) [inline src/DynamoCoreWpf/UI/Converters.cs:3646]: add margin/Padding and return the added value?

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+using System.Runtime.CompilerServices;
+using System.Windows.Media;
+using Dynamo.Core;
+using Xceed.Wpf.Toolkit;
+
+namespace Dynamo.ViewModels
+{
+    internal class CustomColorItem : ColorItem, INotifyPropertyChanged
+    {
+        public event PropertyChangedEventHandler PropertyChanged;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:14]: Comments for public properties please

**Comment by RobertGlobant20** [comment]
> From the GIF I see that there is considerable lag when you open the color picker for the first time, is it because you are in Debug, or is it usual?

@zeusongit 
GIF updated using my new laptop:
AnimationColorPicker

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                return isColorItemSelected;
+            }
+            set
+            {
+                isColorItemSelected = value;
+                OnPropertyChanged(nameof(IsColorItemSelected));
+            }
+        }
+
+        public CustomColorItem(Color? color, string name) : base(color, name)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:30]: If this does nothing, do we need it?

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            CustomColors = new ObservableCollection<CustomColorItem>();
+            CustomColors = CreateCustomColorsCollection();
+        }
+
+        private static ObservableCollection<CustomColorItem> CreateBasicColorsCollection()
+        {
+
+            ObservableCollection<CustomColorItem> observableCollection = new ObservableCollection<CustomColorItem>();
+
+            List<(int R, int G, int B)> colors = new List<(int, int, int)>()
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:112]: Would you add a comment about how the following colors are defined or where (link) to find the definition?

**Review by QilongTang — COMMENTED** [review]
LGTM with comments

**Review by zeusongit — APPROVED** [review]

File: src/DynamoCoreWpf/UI/Converters.cs (hunk 2 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/UI/Converters.cs]: @QilongTang the comment was updated in the next commit: https://github.com/DynamoDS/Dynamo/pull/13794/commits/a49d970a769a23938d84abc67836a231e4de443d

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs (hunk 3 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:17]: @QilongTang comments added in the next commit: https://github.com/DynamoDS/Dynamo/pull/13794/commits/a49d970a769a23938d84abc67836a231e4de443d

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs (hunk 4 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:38]: yes, it initialize the base class ColorItem with the parameters passed, I've added comments about it in the next commit:
https://github.com/DynamoDS/Dynamo/pull/13794/commits/a49d970a769a23938d84abc67836a231e4de443d

File: src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs (hunk 5 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/ViewModels/Core/CustomColorPickerViewModel.cs:121]: @QilongTang extra comments added (let me know if is enough) in the next commit: https://github.com/DynamoDS/Dynamo/pull/13794/commits/a49d970a769a23938d84abc67836a231e4de443d

File: src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs [hunk 6]
```diff
@@ -1079,5 +1079,10 @@ private void OnGeometryScaling_Click(object sender, RoutedEventArgs e)
             }
             GeoScalingPopup.IsOpen = true;
         }
+
+        private void InCanvasSearchBar_Opened(object sender, EventArgs e)
+        {
+            int i = 0;
+        }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs:1086]: This is not meant to be committed right?

**Comment by QilongTang** [comment]
Once the last comment addressed, good to go

File: src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs (hunk 6 above)
Comment (RobertGlobant20) [reply src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs]: @QilongTang code removed in the next commit: https://github.com/DynamoDS/Dynamo/pull/13794/commits/e3416c6dbc972187f66a46f9243770f360d2c144

**Comment by RobertGlobant20** [comment]
> Once the last comment addressed, good to go. @RobertGlobant20 Please also manul test adding group styles locally to see if the data in settings file is in a consistent format.

@QilongTang here is a GIF of how the GroupStyles are serialized in DynamoSettings.xml file.
TestSerializerColorPicker

**Pull request merged**